diff -ru8bBwd spdlog-1.15.0.orig/meson.build spdlog-1.15.0/meson.build
--- spdlog-1.15.0.orig/meson.build	2025-01-23 08:31:58.995381813 +0100
+++ spdlog-1.15.0/meson.build	2025-01-23 08:33:33.285655794 +0100
@@ -11,55 +11,33 @@
 thread_dep = dependency('threads')
 
 spdlog_dependencies = [thread_dep]
 spdlog_compile_args = []
 
 if meson.get_compiler('cpp').has_header_symbol('format', '__cpp_lib_format', required: get_option('std_format'))
   spdlog_compile_args += '-DSPDLOG_USE_STD_FORMAT'
   fmt_dep = dependency('', required: false)
-else
-  fmt_dep = dependency('fmt', version: '>=8.1', required: get_option('external_fmt'))
-  if fmt_dep.found()
-    spdlog_dependencies += fmt_dep
-    spdlog_compile_args += '-DSPDLOG_FMT_EXTERNAL'
-  endif
 endif
 
 if get_option('compile_library')
   spdlog_compile_args += '-DSPDLOG_COMPILED_LIB'
   if get_option('default_library') != 'static'
     spdlog_compile_args += '-DSPDLOG_SHARED_LIB'
     spdlog_compile_args += '-Dspdlog_EXPORTS'
-    if not fmt_dep.found()
       spdlog_compile_args += '-DFMT_EXPORT'
       spdlog_compile_args += '-DFMT_SHARED'
     endif
-  endif
   subdir('src')
 
   spdlog_dep = declare_dependency(
     include_directories: inc,
     dependencies: spdlog_dependencies,
     link_with: spdlog_lib,
     compile_args: spdlog_compile_args,
   )
 else
   spdlog_dep = declare_dependency(
     include_directories: inc,
     dependencies: spdlog_dependencies,
     compile_args: spdlog_compile_args,
   )
 endif
-
-# install header and pkg-config file
-install_subdir('include/spdlog', install_dir: get_option('includedir'))
-pkg = import('pkgconfig')
-pkg.generate(
-  name: 'libspdlog',
-  filebase: 'spdlog',
-  description: 'Fast C++ logging library.',
-  url: 'https://github.com/gabime/spdlog',
-  libraries: [spdlog_dep, fmt_dep],
-  extra_cflags: spdlog_compile_args,
-)
-
-subdir('tests')
diff -ru8bBwd spdlog-1.15.0.orig/src/meson.build spdlog-1.15.0/src/meson.build
--- spdlog-1.15.0.orig/src/meson.build	2025-01-23 08:31:58.995381813 +0100
+++ spdlog-1.15.0/src/meson.build	2025-01-23 08:33:19.400788837 +0100
@@ -1,20 +1,17 @@
 src = files(
   'async.cpp',
   'cfg.cpp',
   'color_sinks.cpp',
   'file_sinks.cpp',
   'spdlog.cpp',
   'stdout_sinks.cpp',
+  'bundled_fmtlib_format.cpp'
 )
 
-if not fmt_dep.found()
-  src += files('bundled_fmtlib_format.cpp')
-endif
-
 spdlog_lib = library(
   'spdlog',
   src,
   include_directories: inc,
   dependencies: spdlog_dependencies,
   cpp_args: spdlog_compile_args,
 )
